### PR TITLE
Mega Update

### PR DESCRIPTION
**Pull request by Jie-F**
@TimArnettThales
This PR is what I hope to be a solid step forward toward your vision of the game, and sets a good foundation for v3. I'm happy to revise or discuss anything, and would appreciate a careful review so that nothing problematic slips in.

@Cedric-Boucher
Thank you for helping brainstorm the interface! I'd appreciate your thoughts on this new implementation.

# Summary:
Perhaps this could be considered the final version of Kessler v2.X.X. While visually and mechanically hard to distinguish from previous versions, under the hood this version is much faster, cleaner, and a more mathematically precise simulation. The core loop uses lots of continuous time physics which are framerate independent, and the controller interface has been updated to be more flexible and performant.

# MyPyC #95 

MyPyC compilation has been fixed and automated! For people on Windows running Python 3.10 up to 3.13, doing `pip install kesslergame` will install and run the compiled wheel starting from this new version.

CONTRIBUTING.md has been updated with a note suggesting developers frequently run `mypy --strict kessler_game.py` during development to statically type check the code, preventing bugs, and allowing MyPyC compilation to complete without error.

# Performance #103 

Compared to the last PyPI version v2.2.0, this new version is over 20X faster! This includes the time for the game, as well as for the controller to parse and read all the data passed in from the game each frame.

This is a benchmark with 30 asteroids, 1200 potential asteroids, 30 FPS. Due to the game version difference, the scenarios lasted a slightly different amount of time, but I did 100 trials each and used a random seed for each trial, so the version differences average out:

|            Game Config             | Average time per trial | Total scenario realtime ratio | Relative Speed to v2.2.0 |
|:---------------------------------:|:-----------------------:|:-----------------------------:|:-------------------------:|
| New .compact representation, MyPyC |       0.02091s       |            5096X            |           33.3×           |
|          v2.2.0 Pure Python               |       0.6621s        |             153X            |           1.0×            |

So this suggests this new version is 33.3X faster than the last version! But this is kind of cheating, because this is just the speed the game runs at. The controller is not reading the gamestate or anything. If we include the time it takes for a C++ controller that uses Nanobind to parse through and read the data, the overall speed drops, along with the overall speed benefit of this new version. However, keep in mind this new version sends the gamestate in a MUCH more efficient format, that is multiple times quicker for the controller to extract the fields from. It's just that overall the controller data reading didn't get as insane of a speedup as the main game does, so it seems like overall the speedup is not as extreme.

|            Game Config             | Average time per trial | Total scenario realtime ratio | Relative Speed to v2.2.0 |
|:---------------------------------:|:-----------------------:|:-----------------------------:|:-------------------------:|
| New .compact representation (comp unsafe) |       0.03520s       |            3028X            |           23.3×           |
| New .compact representation (comp safe)   |       0.04579s       |            2327X            |           17.9×           |
|          v2.2.0 Pure Python               |       0.7785s        |             130X            |           1.0×            |

23.3X faster. Not too shabby. It'll still let you run an hour of the game in just over a second (corresponding to a realtime ratio of 3600X).

Thanks to the new compact internal representation of state objects as mutable lists, turning on competition safe mode only gives a small performance penalty! It used to be much worse, like 5X slower when dicts were still used and had to be copied. Comp safe mode is definitely the default, but people can choose to use unsafe mode for development, to eek out that extra bit of performance.

And the overall breakdown of what time is spent on what, still over half of the time is spent maintaining the gamestate views and having the controller parse through it. A minority of the time is spent on collision checks. So the internal game calculations are not a bottleneck. And even speeding up the game calculations infinitely, will only yield a 2X speed increase overall.

To achieve this speedup, a LOT changed in the game. MyPyC alone is responsible for about 4X of the speedup. The rest is down to more efficient internal representations of the game data, and making things leaner overall. You'll see this in the next sections of changes.

# New GameState Interface #103 

Historically the game_state and ship_states passed to the controllers were pure Python dictionaries. This had the advantage of using primitive Python types and is very simple. However, dictionaries can be slow and heavy to create, mutate, copy, and read from. The game state dictionaries were the bottleneck in the performance of the game, on both the game's end for creating and maintaining them, as well as on the controller's end, parsing through the dictionaries and reading the values each frame. In a simple benchmark of a controller that only reads the data coming in from the game, 90% of the time was spent on passing data back and forth!

There is now a new state_models.py that defines a bunch of new interfaces. This PR proposes changing to passing in a new GameState and ShipState class instance. And with these classes, there are now three interfaces that contestants can choose from, each with their pros and cons.

## Dictionary

The class can be accessed exactly like before, pretending it is a dictionary. This provides backward compatibility and simplicity.

E.g. `game_state["asteroids"]`

## Class Attributes

This provides a very convenient-to-type way to access the game data.

E.g. `game_state.asteroids[4].radius`, `game_state.time_limit`, `game_state.map_size`

## Compact

This representation of data is the core of the new optimization to speed up data mutation and passing. Asteroid game objects are stored the same way as before, but each one holds a reference to a mutable list that stores the state of the object. When objects like asteroids update, they mutate this list to update the position, and in turn this change reflects in the GameState object. And the `game_state.compact` and `ship_state.compact` exposes these lists directly to the controller, so they can choose to parse a list instead of parsing a dictionary to get the info behind a game object.

This is the fastest way for a controller to access and read the gamestate with minimal overhead, and is recommended for advanced users who want the fastest training times, or have expensive computations occurring within their agent, where every millisecond matters.

An example of the dictionary the .compact property returns is:

```
{
    "ships": [
        [385.0607585391769, 747.9475394887982, -4.408728476930471e-14, -240.0, 240.0, 270.0, 300.0, 20.0, 1, 1, False, 3, 0]
    ],
    "asteroids": [
        [600.0000000000009, 233.673456, 100.0, -2.4492935982947064e-14, 3, 452.3893421169302, 24.0],
        [500.0, 33.67345800000012, -3.6739403974420595e-14, -200.0, 1, 50.26548245743669, 8.0],
        [500.0, 733.773458, -5.510910596163089e-14, -300.0, 4, 804.247719318987, 32.0],
    ],
    "bullets": [
        [279.15311794985496, 681.3712368493217, -247.21359549995788, 760.8452130361229, 3.708203932499368, -11.412678195541844, 108.0, 1.0, 12.0],
        [42.00744880738853, 340.4369113621759, -647.2135954999578, 470.2282018339786, 9.708203932499368, -7.0534230275096785, 144.0, 1.0, 12.0],
        [66.13819320973143, 24.133503779823066, -800.0, 9.797174393178826e-14, 12.0, -1.4695761589768238e-15, 180.0, 1.0, 12.0],
    ],
    "mines": [],
    "map_size": (1000, 800),
    "time_limit": inf,
    "time": 0.9999999999999999,
    "frame": 30,
    "delta_time": 0.03333333333333333,
    "frame_rate": 30.0,
    "random_asteroid_splits": False,
    "competition_safe_mode": True,
}
```

### Minor Breaking Change
Before, the simulation time attributes for seconds and frames were called `time` and `sim_frame`, respectively. For consistency, I renamed `sim_frame` to frame in the external interface. If we would rather have these both be called `sim_time` and `sim_frame` then we can also change it to that or something else. I'm suggesting taking out the sim prefix for simplicity, so it's easier for people to remember what it's called, and people won't forget and think it's `frame_sim` or something else

# Competition Safe Mode
This is a switch defaulting to True. When True, each frame the game makes a copy of the gamestate for each controller, so that no matter how they mutate it, it won't affect the gamestate view for the next frame, or for the next controller.

When set to False, the game will assume that people are careful to not mutate this, and that people will make copies of the data they intend to keep. This will also not guarantee the game_state and ship_state stay the same between frames, since it's the same mutable reference shared by the game and the controller for the entire game! This mode set to False is about 30% faster than True. In cases where the controller is making a copy of the data anyway, it's safe to set this to False and get a free 30% speed boost for training!

# Documentation

The GameState and ShipState interfaces have documentation in docs/ and people can get a list of everything they have access to. This will be helpful to refer to while making an agent, otherwise it's easy to forget what the field names are, and have to read the game's source code or dump out game_state or ship_state to check.

# Mechanical Changes

## Map edge oscillation #101 

Previously if you put an asteroid at the very edge of a map with just the right speed, it would oscillate and wrap every single frame. This has been fixed with a preprocessing step in scenario.py which nudges asteroids away from the edge by an epsilon.

## Framerate-Independent Ship Dynamics #105 

The ship used to move with discrete dynamics, discretely integrating its position each frame t. As delta_time decreases, this approximation gets better and better. But overall, this is framerate dependent. If you lower the framerate down to 5 FPS, you find that the ship accelerates a lot quicker than it should. And turning the ship also behaves differently. It makes it hard to mathematically analyze the path the ship takes.

However if we perform continuous integration assuming the ship's thrust and turn_rate are constant over this frame, then we can make the ship's dynamics mathematically perfect, and framerate independent. The ship's dynamics can then be analyzed using regular infinitesimal calculus, instead of discrete calculus! People with a calculus background and understand integration by parts or Taylor polynomials will be able to replicate this analytic formulation for the ship movement.

## More Fair Collision Handling #104 #100 #99 

Previously if two ships were fighting for the same asteroid, if two bullets hit the same asteroid in the same frame, then whoever Ship/Team 1 is would always get the hit since their bullet was in the list earlier. This has been replaced with a global chronological collision resolution system, where the continuous collision detection time is used as the tiebreaker.

Similarly, what if in the rare chance that two mines blew up at the same time, and an asteroid is within the radius of both of them? Previously, Ship 1's mine gets the credit assigned for the hit. Now, whichever mine was closer to the asteroid gets the credit assigned.

# Removing src module

@TimArnettThales I would really appreciate if we can consider removing the src module which causes many issues. Switching away from it is as easy as changing a setting in your IDE, and removing the "src." prefix from imports.

With the MyPyC version of the game installed globally, that can only be accessed by importing "kesslergame". The src. is only a shortcut to import just the game within the repo, and not the global package.

Me and some other friends have had issues with the src. import, and the solution had usually been to remove it and import the global kesslergame module. I think removing this module duplication along with documenting some recommendations could make things less confusing and help solve those issues as well.

Especially with the new MyPyC stuff, I don't think there's a way around removing the src module. It is nonstandard practice, and causes a lot of issues with module duplication. I've spent at least an hour trying to fix it and can't without major compromises on the type checking front. This is because with MyPyC runtime type-checked code, if any scenarios are created from the Scenario class imported using `from src.kesslergame import Scenario` but then run the game with the "kesslergame" compiled module, this error will occur:

`TypeError: kesslergame.scenario.Scenario object expected; got src.kesslergame.scenario.Scenario`

These two modules are distinct, because the src module is different than the kesslergame module.

The good news is that fixing this is very easy, and we can document this :)

Within PyCharm, please first try removing the "src." prefix from where the Scenario import is happening, so that it tries to imports the Scenario object from the correct kesslergame package instead of the src package. If you then get this error:
image

Right click the src folder -> Mark directory as -> Sources root
image

In VS Code, I think the solution is creating or editing `.vscode/settings.json` and putting in:
`{"python.analysis.extraPaths": ["src"]}`
And this should tell VS Code to look in src when searching for imports, and the kesslergame package will be found.

Alternatively, if you just want to run the game and don't want to modify it, if you `pip install kesslergame` globally or within the venv, that will make it importable.

And that should allow kesslergame to be detected as the package. I think this is the standard way to set things up, and it will resolve this module duplication error, and fix runtime type checking along with MyPyC wheel compilation without any workarounds. I also hope that this will make it easier for newcomers to avoid module import errors too, and reduce confusion there.

# Notes for Review

I think it's a good idea to not be in a rush to merge this PR so there's plenty of time to review, discuss, and revise. Please feel free to leave comments and questions in the code or in general.

And also this will give me more time to test. I need to update my controller to work with this version and have the time to fully test things, and maybe find some edge cases. And maybe along the way I'll add more documentation or little formatting changes to the code to polish things off.

Also, what version number should this be? I added a version reference in the documentation, so I’d like to update that to match.

### CODE REVIEW

File: src/kesslergame/ship.py [hunk 1]
```diff
@@ -196,45 +239,146 @@ def update(self, delta_time: float = 1 / 30) -> tuple[Optional[Bullet], Optional
...
+                vi: initial speed
+                a: acceleration
+                th0: initial heading (radians)
+                omega: turn rate (rad/sec)
+                t: duration (seconds)
+            """
+            if abs(t) < 1e-12:
+                return 0.0, 0.0
+            if abs(omega) < 1e-4:
+                # Taylor expansion with Horner's method
```
Comment (Jie-F) [inline src/kesslergame/ship.py:300]: K this is totally wrong. I'll fix this once I do the math, but it's actually pretty tricky

**Comment by Jie-F** [comment]
I'm gonna see how feasible a continuous ship-asteroid and ship-ship collision is to do, now that the movement is mathematically exact. It'll probably end up being a transcendental equation that needs a root finder. Best case, a Weierstrass sub turns it into a quartic polynomial or something.

It won't really slow down the game since each frame, you only need to do the check for objects nearby, and there might only be a couple checks each frame. So use a broad phase check before the narrow phase check.

**Comment by Jie-F** [comment]
All collisions in the game are now continuously calculated, and chronologically resolved! I did some basic testing and it seems to work well. More thorough testing will be to follow. And perhaps more important than 100% correctness, is that the code never crashes due to some pathological divide by zero 😅

**Comment by Jie-F** [comment]
Oop almost forgot. I'd also like to change the ship update order from "shoot then move", to "move then fire". This ALWAYS trips up newbies, and every time I get someone into Kessler and ask them whether they think the ship should fire from its new heading/location or old, they instinctively say the new heading. But that's not how the game works! And I explain to them that to get a shot to line up, you need to actually wait an extra frame for the ship to turn all the way, and use a boolean flag to indicate that you should shoot the next frame, so you can begin turning toward your next target and remember you wanted to shoot. So anecdotally, this old behavior is counterintuitive. It makes code clunky, and you have to maintain firing state across frames.

Also, the control action tuple is:
thrust, turn_rate, fire, drop mine.
Which once again primes the brain to think that the game first moves, then fires! So we should align with that.

First move, then fire, and then resolve collisions with the projectiles that came out of the ship's new position. In a continuous simulation, this is the only way that makes sense. Otherwise, the bullet is one frame from the past, yet it can still collide with objects in the present.

**Comment by Jie-F** [comment]
Also I think we should add a frame_skip game option which makes it so the game only draws/updates a frame if frame % frame_skip == 0. The Tkinter graphics is super slow and slows down the game even if the realtime multiplier is up higher. So with a frame_skip param that defaults to something like floor(realtime_multiplier), this will make the game run a LOT faster without graphics trying to draw each frame and struggle to keep up. This will help a lot with testing agents with visuals where you want to fast forward without disabling graphics altogether.

**Comment by TimArnettThales** [comment]
Frame skip idea and move first changes are more than fine with me. Also sounds good re: src module and all of the other updates with continuous representations of collisions and movement integration.

**Comment by TimArnettThales** [comment]
Am I understanding the performance table correctly for #103 ? It appears that mypyc isn't improving the speed that much compared to just the new compact game state. It's obviously still beneficial to have but just curious if that's correct that compact unsafe is 23.3x and mypyc (I assume also unsafe) is 33.3x.

**Comment by Jie-F** [comment]
And remember that MyPyC can only speed up the Kessler side of things, while the benchmark also includes the time it takes for the controller to read the data. So Kessler could be 8X faster due to compilation, but if things were 50/50 split in time before, overall Kessler + controller data read would only be faster by a factor of 100/(50/8+50)=1.78X

**Comment by Jie-F** [comment]
I'd say this would be ready to merge beginning of next week.

The rate of me spotting issues is finally slowing down. There are still some subtle issues left with the continuous collision detection not considering wrap and back-extrapolating objects beyond the border instead of reporting no collision, but those are occasional edge cases. I'll still try to fix them for complete fairness.

**Comment by Jie-F** [comment]
I'll probably refactor the ship integration stuff into smaller helpers and use that to find all these time intervals using even more root finding, yay!

Also I just realized my ship-ast and ship-ship early rejection code is wrong since I use the ship speed which might have been higher during the frame. abs(speed)+(480+80)*dt would be a more appropriate upper bound.

**Comment by Jie-F** [comment]
To have this in a stable state for release, I've began working on a different branch which will fix the remaining issues and enhance collisions.

**Review by TimArnettThales — APPROVED** [review]
